// module_service.proto

syntax = "proto3";

service ModuleService {
    rpc get_attribute(AttributeRequest) returns (ValueResponse) {}
    rpc execute(MethodRequest) returns (ValueResponse) {}
}

message AttributeRequest {
    string id = 1;
    string attribute = 2;
}

message MethodArgument {
    bool proxy = 1;
    bytes bytes_value = 2;
    string string_value = 3;
}

message MethodRequest {
    string id = 1;
    string method = 2;
    repeated MethodArgument args = 3;
    map<string, MethodArgument> kwargs = 4;
}

message ValueResponse {
    int32 type = 1; // -1 primitive, 0 proxy value, 1 proxy property, 2 callable
    bytes bytes_value = 2;
    string string_value = 3;
}
